### DISCUSSION TITLE
How to add a custom response header?

### DISCUSSION THREAD

**Discussion by williambarberjr** [hasura/graphql-engine#7445] [Question] [answered]
How can I add a custom response header that returns the total number of rows from the database that match the query? 

So if 10 rows came back from a query, I'd want a response header that says:
"Custom-API-Response-Header": Rows=10

**Comment by leoalves** [comment, marked as answer]
To get the total number of rows you can use the aggregate queries from Hasura.
https://hasura.io/docs/latest/graphql/core/databases/postgres/queries/aggregation-queries.html

But it is not possible to add a header to the response.

**Reply by williambarberjr** [reply]
Bummer, I assume I can't add a response header even if I define my own rest api endpoint using hasura?

**Reply by motleydev** [reply]
Out of curiosity, what's the problem you are trying to solve for? To get custom headers, you would indeed need to write a proxy service wrapping the response.

**Reply by williambarberjr** [reply]
I'm looking for the simplest (lowest-code) way to monetize access to a large dataset via a paid API. In this case via rapidapi's marketplace. 

While I could do this with hasura if I setup a rest endpoint to only allow each request to return one row of the database, if I'd like the user to be able to request hundreds of rows without sending hundreds of individual requests, then rapidapi requires my api to respond with a custom response header that specifies how many units the request should be charged for. Without that header, a user could request the entire database and be charged pennies because it's billed per request by default.

**Reply by williambarberjr** [reply]
What about sending the total number of rows returned to another api endpoint each time one of the hasura endpoints is used (per the description here: https://docs.blobr.io/how-to-use-blobr/2.-create-a-plan/untitled/based-on-custom-metric). Can the "actions" functionality with custom business logic handle this?

**Reply by BenoitRanque** [reply]
You could look into deriving actions for this purpose.
Your action endpoint would then forward the query back to hasura, get and measure the result, notify an external api about the size of the result, and finally forward that result to the client. This implies creating an action for each monetized query, which may be a lot of work.

Alternatively, you could setup a proxy that simply measures the size of the response in kilobytes. This is assuming that size in kilobytes correlates to the billable amount.

**Comment by thefedoration** [comment]
+1 to adding custom response headers. In my case it's adding security headers like the ones assessed on securityheaders.com
<img width="1273" alt="Screenshot 2023-02-21 at 3 52 55 PM" src="https://user-images.githubusercontent.com/6766645/220485264-ec578c7f-204f-4ff7-9152-639fba6472f7.png">
